File: Frontend/src/assets/fable_build/Engine/StrategyEvaluator.js
import { length, singleton, append, empty, sumBy, isEmpty, partition } from "../fable_modules/fable-library-js.4.27.0/List.js";
import { price } from "../Simulation/PricingModels.js";
import { SimulationRunResult, EvaluationState, Transaction, Portfolio } from "./EngineTypes.js";
import { calculatePortfolioValue } from "./PortfolioQueries.js";
import { reconcileCash } from "./Reconciler.js";
import { interpretStep, emptyState } from "./Interpreter.js";
import { setItem } from "../fable_modules/fable-library-js.4.27.0/Array.js";

function processSettlement(portfolio, history, currentDay, riskFreeRate) {
    const patternInput = partition((p) => {
        const matchValue = p.Instrument;
        if (matchValue.tag === 1) {
            return matchValue.fields[0].ExpiryDay <= currentDay;
        }
        else {
            return false;
        }
    }, portfolio.Positions);
    const expired = patternInput[0];
    if (isEmpty(expired)) {
        return portfolio;
    }
    else {
        return new Portfolio(portfolio.Cash + sumBy((p_1) => {
            const matchValue_1 = p_1.Instrument;
            if (matchValue_1.tag === 1) {
                return (price(matchValue_1.fields[0], history, currentDay, riskFreeRate) * p_1.Quantity) * 100;
            }
            else {
                return 0;
            }
        }, expired, {
            GetZero: () => 0,
            Add: (x, y) => (x + y),
        }), patternInput[1], portfolio.CompositeRegistry, portfolio.TaxLots, portfolio.TaxLiabilityYTD, portfolio.RealizedGainsYTD);
    }
}

function processCashflows(portfolio, scenario, currentDay, history, riskFreeRate, costs, taxConfig) {
    let matchValue;
    let currentPortfolio = portfolio;
    let transactions = empty();
    if ((matchValue = taxConfig.PaymentMode, (matchValue.tag === 1) ? ((currentDay > 0) && ((currentDay % matchValue.fields[0]) === 0)) : ((currentDay > 0) && ((currentDay % 252) === 0)))) {
        const totalTax = currentPortfolio.TaxLiabilityYTD + (calculatePortfolioValue(currentPortfolio, history, currentDay, riskFreeRate) * taxConfig.WealthTaxRate);
        if (totalTax > 0) {
            const patternInput = reconcileCash(currentPortfolio, totalTax, history, currentDay, riskFreeRate, costs);
            const postTaxPortfolio = patternInput[0];
            const taxTxn = new Transaction(currentDay, "TAX_PAYMENT", "TAX", 0, 0, -totalTax, 0, 0, totalTax, "PERIODIC_SETTLEMENT");
            currentPortfolio = (new Portfolio(postTaxPortfolio.Cash, postTaxPortfolio.Positions, postTaxPortfolio.CompositeRegistry, postTaxPortfolio.TaxLots, 0, postTaxPortfolio.RealizedGainsYTD));
            transactions = append(transactions, append(patternInput[1], singleton(taxTxn)));
        }
    }
    if ((currentDay > 0) && ((currentDay % 30) === 0)) {
        const year = currentDay / 365;
        const currentMonth = ~~(currentDay / 30) | 0;
        switch (scenario.tag) {
            case 1: {
                const p = scenario.fields[0];
                const amount = p.MonthlyContribution * Math.pow(1 + p.ContributionGrowthRate, Math.floor(year));
                currentPortfolio = (new Portfolio(currentPortfolio.Cash + amount, currentPortfolio.Positions, currentPortfolio.CompositeRegistry, currentPortfolio.TaxLots, currentPortfolio.TaxLiabilityYTD, currentPortfolio.RealizedGainsYTD));
                break;
            }
            case 2: {
                const p_1 = scenario.fields[0];
                const inflationFactor = Math.pow(1 + p_1.InflationRate, Math.floor(year));
                const netWithdrawal = (p_1.MonthlyWithdrawal * inflationFactor) - ((currentMonth >= p_1.PensionStartMonth) ? (p_1.MonthlyPension * inflationFactor) : 0);
                if (netWithdrawal <= 0) {
                    currentPortfolio = (new Portfolio(currentPortfolio.Cash + Math.abs(netWithdrawal), currentPortfolio.Positions, currentPortfolio.CompositeRegistry, currentPortfolio.TaxLots, currentPortfolio.TaxLiabilityYTD, currentPortfolio.RealizedGainsYTD));
                }
                else if (currentPortfolio.Cash >= netWithdrawal) {
                    currentPortfolio = (new Portfolio(currentPortfolio.Cash - netWithdrawal, currentPortfolio.Positions, currentPortfolio.CompositeRegistry, currentPortfolio.TaxLots, currentPortfolio.TaxLiabilityYTD, currentPortfolio.RealizedGainsYTD));
                }
                else {
                    const patternInput_1 = reconcileCash(currentPortfolio, netWithdrawal, history, currentDay, riskFreeRate, costs);
                    currentPortfolio = patternInput_1[0];
                    transactions = append(transactions, patternInput_1[1]);
                }
                break;
            }
            default:
                undefined;
        }
    }
    return [currentPortfolio, transactions];
}

export function evaluate(runId, program, config, history, initialCash) {
    let matchValue, TransactionHistory;
    let currentState = emptyState((matchValue = config.Scenario, (matchValue.tag === 2) ? matchValue.fields[0].InitialPortfolio : initialCash), config.RiskFreeRate);
    const equityCurve = new Float64Array(config.TradingDays + 1);
    for (let day = 0; day <= config.TradingDays; day++) {
        currentState = (new EvaluationState(day, currentState.Portfolio, currentState.ScopeStack, currentState.GlobalScope, currentState.RiskFreeRate, currentState.TransactionHistory));
        const settledPortfolio = processSettlement(currentState.Portfolio, history, day, config.RiskFreeRate);
        currentState = (new EvaluationState(currentState.CurrentDay, settledPortfolio, currentState.ScopeStack, currentState.GlobalScope, currentState.RiskFreeRate, currentState.TransactionHistory));
        const patternInput = processCashflows(currentState.Portfolio, config.Scenario, day, history, config.RiskFreeRate, config.ExecutionCosts, config.Tax);
        currentState = ((TransactionHistory = append(currentState.TransactionHistory, patternInput[1]), new EvaluationState(currentState.CurrentDay, patternInput[0], currentState.ScopeStack, currentState.GlobalScope, currentState.RiskFreeRate, TransactionHistory)));
        if (((length(currentState.Portfolio.Positions) > 0) ? true : (currentState.Portfolio.Cash > 0)) && ((day === 0) ? true : ((day % config.Granularity) === 0))) {
            currentState = interpretStep(program, currentState, history, config.ExecutionCosts, config.Tax);
        }
        const dailyValue = calculatePortfolioValue(currentState.Portfolio, history, day, config.RiskFreeRate);
        setItem(equityCurve, day, dailyValue);
    }
    return new SimulationRunResult(runId, equityCurve, currentState, currentState.TransactionHistory);
}

File: Backend/StrategyEngine/Engine/StrategyEvaluator.fs
module StrategyEvaluator

open System
open AST
open EngineTypes
open Interpreter
open PortfolioQueries
open PricingModels
open StrategyEngine.Engine.Reconciler

let private processSettlement (portfolio: Portfolio) (history: FullPriceHistory) (currentDay: int) (riskFreeRate: float) : Portfolio =
    let (expired, active) = 
        portfolio.Positions 
        |> List.partition (fun p -> 
            match p.Instrument with
            | ResolvedOption opt -> opt.ExpiryDay <= currentDay
            | _ -> false
        )

    if expired.IsEmpty then 
        portfolio
    else
        let settlementCash = 
            expired 
            |> List.sumBy (fun p -> 
                match p.Instrument with
                | ResolvedOption opt -> 
                    let unitValue = price opt history currentDay riskFreeRate
                    unitValue * p.Quantity * 100.0
                | _ -> 0.0
            )

        { portfolio with 
            Cash = portfolio.Cash + settlementCash
            Positions = active 
        }

let private processCashflows 
    (portfolio: Portfolio) 
    (scenario: FinancialScenario) 
    (currentDay: int) 
    (history: FullPriceHistory) 
    (riskFreeRate: float) 
    (costs: ExecutionCosts)
    (taxConfig: TaxConfig)
    : Portfolio * Transaction list =
    
    let mutable currentPortfolio = portfolio
    let mutable transactions = []

    let isTaxDay = 
        match taxConfig.PaymentMode with
        | ImmediateWithholding -> currentDay > 0 && currentDay % 252 = 0 
        | PeriodicSettlement freq -> currentDay > 0 && currentDay % freq = 0

    if isTaxDay then
        let capGainsTax = currentPortfolio.TaxLiabilityYTD
        let nav = calculatePortfolioValue currentPortfolio history currentDay riskFreeRate
        let wealthTax = nav * taxConfig.WealthTaxRate
        let totalTax = capGainsTax + wealthTax

        if totalTax > 0.0 then
            let (postTaxPortfolio, liqTxns) = reconcileCash currentPortfolio totalTax history currentDay riskFreeRate costs
            
            let taxTxn = {
                Date = currentDay
                Ticker = "TAX_PAYMENT"
                Type = "TAX"
                Quantity = 0.0
                Price = 0.0
                Value = -totalTax
                Commission = 0.0
                Slippage = 0.0
                Tax = totalTax
                Tag = Some "PERIODIC_SETTLEMENT"
            }
            
            currentPortfolio <- { postTaxPortfolio with TaxLiabilityYTD = 0.0 }
            transactions <- transactions @ liqTxns @ [taxTxn]

    if currentDay > 0 && currentDay % 30 = 0 then
        let year = float currentDay / 365.0
        let currentMonth = currentDay / 30
        
        match scenario with
        | NoScenario -> ()
        
        | Accumulation p ->
            let growthFactor = Math.Pow(1.0 + p.ContributionGrowthRate, Math.Floor(year))
            let amount = p.MonthlyContribution * growthFactor
            currentPortfolio <- { currentPortfolio with Cash = currentPortfolio.Cash + amount }
            
        | Retirement p ->
            let inflationFactor = Math.Pow(1.0 + p.InflationRate, Math.Floor(year))
            let grossExpenses = p.MonthlyWithdrawal * inflationFactor
            
            let pensionIncome = 
                if currentMonth >= p.PensionStartMonth then 
                    p.MonthlyPension * inflationFactor
                else 
                    0.0
            
            let netWithdrawal = grossExpenses - pensionIncome
            
            if netWithdrawal <= 0.0 then
                currentPortfolio <- { currentPortfolio with Cash = currentPortfolio.Cash + abs(netWithdrawal) }
            else
                if currentPortfolio.Cash >= netWithdrawal then
                    currentPortfolio <- { currentPortfolio with Cash = currentPortfolio.Cash - netWithdrawal }
                else
                    let (postWithdrawal, liqTxns) = reconcileCash currentPortfolio netWithdrawal history currentDay riskFreeRate costs
                    currentPortfolio <- postWithdrawal
                    transactions <- transactions @ liqTxns

    (currentPortfolio, transactions)

let evaluate 
    (runId: int)
    (program: Program)
    (config: SimulationConfiguration)
    (history: FullPriceHistory)
    (initialCash: float)
    : SimulationRunResult =

    let startCash = 
        match config.Scenario with
        | Retirement p -> p.InitialPortfolio
        | _ -> initialCash

    let initialState = Interpreter.emptyState startCash config.RiskFreeRate
    
    let mutable currentState = initialState
    let equityCurve = Array.zeroCreate (config.TradingDays + 1)

    for day in 0 .. config.TradingDays do
        
        currentState <- { currentState with CurrentDay = day }

        let settledPortfolio = processSettlement currentState.Portfolio history day config.RiskFreeRate
        currentState <- { currentState with Portfolio = settledPortfolio }

        let (cashflowPortfolio, liqTransactions) = 
            processCashflows currentState.Portfolio config.Scenario day history config.RiskFreeRate config.ExecutionCosts config.Tax
            
        currentState <- { 
            currentState with 
                Portfolio = cashflowPortfolio
                TransactionHistory = currentState.TransactionHistory @ liqTransactions 
        }

        let isAlive = currentState.Portfolio.Positions.Length > 0 || currentState.Portfolio.Cash > 0.0
        
        if isAlive && (day = 0 || day % config.Granularity = 0) then
            currentState <- Interpreter.interpretStep program currentState history config.ExecutionCosts config.Tax

        let dailyValue = PortfolioQueries.calculatePortfolioValue currentState.Portfolio history day config.RiskFreeRate
        equityCurve.[day] <- dailyValue

    {
        RunId = runId
        EquityCurve = equityCurve
        FinalState = currentState
        TransactionHistory = currentState.TransactionHistory
    }